### PR TITLE
[@types/node]: Add argument types for `net.Server.close` callback

### PR DESCRIPTION
**Pull request by bhongy**
**Changing an existing definition:**

Replace the optional callback type of `net.Server.close` from `Function` to `(err?: Error) => void`. The callback might be called with `undefined` or an error.

This change will allow users to leverage on the library type instead of having to provide the error type:

From:
```typescript
server.close((err: Error) => (err ? reject(err) : resolve()));
```

To:
```typescript
// type of `err` is provided by the libdef
server.close(err => (err ? reject(err) : resolve()));
```

Reference:
[node v10.15.1 source code](https://github.com/nodejs/node/blob/297c9d1aa9dd97692a33e9c1e863ea9d053c6d58/lib/net.js#L1559)
[node v8.15.0 source code](https://github.com/nodejs/node/blob/6d1c367cf1ef1ab49302e0745dfaef1cdea37d44/lib/net.js#L1609)
[node v6.16.0 source code](https://github.com/nodejs/node/blob/a3aa4b41fd1aea4649d16971a52616b4c6c3af37/lib/net.js#L1513)

Note: I update (11—current, 10, 9, 8) due to the [LTS schedule](https://nodejs.org/en/about/releases/) supports v6 until April 2019 (and I use v8+). Although, happy to help updating them (7 and 6) if need to.

Please advise if I need to do version bump manually.

---
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `yarn lint node`

### CODE REVIEW

**Comment by bhongy** [comment]
CI seems to be failing on unrelated tests like `consul`, `adone`, `stella`, `yargs`, etc.

File: types/node/index.d.ts [hunk 1]
```diff
@@ -35,6 +35,7 @@
 //                 Jeremie Rodriguez <https://github.com/jeremiergz>
 //                 Samuel Ainsworth <https://github.com/samuela>
 //                 Kyle Uehlein <https://github.com/kuehlein>
+//                 Thanik Bhongbhibhat <https://github.com/bhongy>
```
Comment (bhongy) [inline types/node/index.d.ts]: Seems like this section used to be alphabetical. Happy to put my name in alphabetically instead.

File: types/node/node-tests.ts [hunk 2]
```diff
@@ -976,10 +976,11 @@ async function asyncStreamPipelineFinished() {
         // close callback parameter is optional
         _server = _server.close();
 
-        // close callback parameter doesn't specify any arguments, so any
-        // function is acceptable
+        // close callback parameter can be either nothing (undefined) or an error
         _server = _server.close(() => { });
-        _server = _server.close((...args: any[]) => { });
+        _server = _server.close((err) => {
+            if (err != null) { const _err: Error = err; }
```
Comment (Flarna) [inline types/node/node-tests.ts]: According to type definiton `err` is either an `Error` or `undefined` therefore a check against `null` seems to be wrong here

File: types/node/test/net.ts [hunk 3]
```diff
@@ -26,9 +26,11 @@ import { LookupOneOptions } from 'dns';
         ipv6Only: true,
     });
 
-    // close has an optional callback function. No callback parameters are
-    // specified, so any callback function is permissible.
-    server = server.close((...args: any[]) => { });
+    // close callback parameter can be either nothing (undefined) or an error
+    server = server.close(() => { });
+    server = server.close((err) => {
+        if (err != null) { const _err: Error = err; }
```
Comment (Flarna) [inline types/node/test/net.ts]: as above

File: types/node/node-tests.ts (hunk 2 above)
Comment (bhongy) [reply types/node/node-tests.ts]: Good point @Flarna. `!= null` is generally used to check not `null` or `undefined` (but not other falsy values). I agree that it's not precise. Thanks for pointing it out.

Addressed in 4809757.

File: types/node/test/net.ts (hunk 3 above)
Comment (bhongy) [reply types/node/test/net.ts]: replied: discussion_r259646643

**Review by Flarna — APPROVED** [review]

File: types/node/index.d.ts (hunk 1 above)
Comment (SimonSchick) [reply types/node/index.d.ts]: We're going to sort it eventually 🙂

**Review by SimonSchick — APPROVED** [review]

**Review by BrunoScheufler — APPROVED** [review]

**Comment by bhongy** [comment]
Resolved merge conflict.

**Review by SimonSchick — APPROVED** [review]

**Pull request merged**